crates/ruff_linter/src/rules/pylint/rules/unnecessary_chained_comparison.rs

```diff
@@ -0,0 +1,114 @@
+use std::collections::{HashMap, HashSet};
+
+use crate::checkers::ast::Checker;
+use ruff_diagnostics::{Diagnostic, Violation};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_ast::{self as ast};
+
+/// ## What it does
+/// Checks for boolean operations such as `a < b and b < c`
+/// that can be refactored into a single comparison `a < b < c`.
+///
+/// ## Why is this bad?
+/// A single comparison is semantically clearer and more concise.
+///
+/// ## Example
+/// ```python
+/// a = int(input())
+/// b = int(input())
+/// c = int(input())
+/// if a < b and b < c:
+///     pass
+/// ```
+///
+/// Use instead:
+/// ```python
+/// a = int(input())
+/// b = int(input())
+/// c = int(input())
+/// if a < b < c:
+///     pass
+/// ```
+
+#[violation]
+pub struct UnnecessaryChainedComparison;
+
+impl Violation for UnnecessaryChainedComparison {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("Simplified chain comparison exists between the operands.")
+    }
+}
+
+// Each integer is a unique identifier for the node.
+#[derive(Default)]
+struct Bounds {
+    lower_bound: HashSet<i32>,
+    upper_bound: HashSet<i32>,
+}
+
+fn update_bounds(
+    operator: ast::CmpOp,
+    id: String,
+    node_id: i32,
+    is_left: bool,
+    uses: &mut HashMap<String, Bounds>,
+) {
+    match operator {
+        ast::CmpOp::Lt | ast::CmpOp::LtE if is_left => {
+            uses.entry(id).or_default().lower_bound.insert(node_id);
+        }
+        ast::CmpOp::Gt | ast::CmpOp::GtE if is_left => {
+            uses.entry(id).or_default().upper_bound.insert(node_id);
+        }
+        ast::CmpOp::Lt | ast::CmpOp::LtE if !is_left => {
+            uses.entry(id).or_default().upper_bound.insert(node_id);
+        }
+        ast::CmpOp::Gt | ast::CmpOp::GtE if !is_left => {
+            uses.entry(id).or_default().lower_bound.insert(node_id);
+        }
+        _ => {}
+    }
+}
+
+fn set_lower_upper_bounds(node: &ast::ExprCompare, uses: &mut HashMap<String, Bounds>) {
+    let mut left_operand: &ast::Expr = &node.left;
+    let node_id = node as *const _ as i32;
+    for (right_operand, operator) in node.comparators.iter().zip(node.ops.iter()) {
+        if let Some(left_name_expr) = left_operand.as_name_expr() {
+            update_bounds(*operator, left_name_expr.id.clone(), node_id, true, uses);
+        }
+
+        if let Some(right_name_expr) = right_operand.as_name_expr() {
+            update_bounds(*operator, right_name_expr.id.clone(), node_id, false, uses);
+        }
+
+        left_operand = right_operand;
+    }
```

Comment:
I don't think this is correct and I'm surprised that Clippy doesn't complain about a possible truncation. 

* `node` is a pointer. The size of a pointer depends on the platform but is 8 bytes large on 64bit systems
* `i32` is 4 bytes large

Maybe @BurntSushi can enlighten me.

Anyway. I think we can get away with something simpler. We don't need globally unique node ids. Instead, we can just use an increasing index:

```patch
Subject: [PATCH] Don't lazily infer imports because it adds an AST dependency to the module looking up the definition
---
Index: crates/ruff_linter/src/rules/pylint/rules/unnecessary_chained_comparison.rs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/crates/ruff_linter/src/rules/pylint/rules/unnecessary_chained_comparison.rs b/crates/ruff_linter/src/rules/pylint/rules/unnecessary_chained_comparison.rs
--- a/crates/ruff_linter/src/rules/pylint/rules/unnecessary_chained_comparison.rs	(revision dbfcdefddeea5496c84530684429019fba4bff87)
+++ b/crates/ruff_linter/src/rules/pylint/rules/unnecessary_chained_comparison.rs	(date 1718868630142)
@@ -43,14 +43,14 @@
 // Each integer is a unique identifier for the node.
 #[derive(Default)]
 struct Bounds {
-    lower_bound: HashSet<i32>,
-    upper_bound: HashSet<i32>,
+    lower_bound: HashSet<u32>,
+    upper_bound: HashSet<u32>,
 }
 
 fn update_bounds(
     operator: ast::CmpOp,
     id: String,
-    node_id: i32,
+    node_id: u32,
     is_left: bool,
     uses: &mut HashMap<String, Bounds>,
 ) {
@@ -71,9 +71,13 @@
     }
 }
 
-fn set_lower_upper_bounds(node: &ast::ExprCompare, uses: &mut HashMap<String, Bounds>) {
+fn set_lower_upper_bounds(
+    node: &ast::ExprCompare,
+    node_id: u32,
+    uses: &mut HashMap<String, Bounds>,
+) {
     let mut left_operand: &ast::Expr = &node.left;
-    let node_id = node as *const _ as i32;
+
     for (right_operand, operator) in node.comparators.iter().zip(node.ops.iter()) {
         if let Some(left_name_expr) = left_operand.as_name_expr() {
             update_bounds(*operator, left_name_expr.id.clone(), node_id, true, uses);
@@ -96,11 +100,13 @@
     }
 
     let mut uses: HashMap<String, Bounds> = HashMap::new();
+    let mut node_id = 0u32;
     for expr in values {
         let Some(compare_expr) = expr.as_compare_expr() else {
             return;
         };
-        set_lower_upper_bounds(compare_expr, &mut uses);
+        set_lower_upper_bounds(compare_expr, node_id, &mut uses);
+        node_id += 1;
     }
 
     for bound in uses.values() {
```

Reply:
Yeah, it definitely could be a truncation here. I would in general be very careful about casting pointers to integers. I like your diff @MichaReiser. :)

Reply:
Great catch, no reason not to use an index here: https://github.com/astral-sh/ruff/pull/11807/commits/6ee4118a9fc3419069b729bff50fd9bb18059f46